### PR TITLE
Dummy sync implementation for no_std

### PR DESCRIPTION
**Pull request by devrandom**
We would like to not bother synchronizing access on no_std platforms, because there is no mature rust multi-threading support for them yet.  Accordingly, this PR implements drop-in replacements for `Mutex`, `Condvar` and (not done yet) `RwLock`.

See also discussion in https://github.com/rust-bitcoin/rust-lightning/pull/842#issuecomment-831645613 - but this PR uses a slightly different approach.

### CODE REVIEW

**Review by TheBlueMatt — APPROVED** [review]
Cool! Glad it kinda just worked. CI should also at least run `cargo test --no-default-features --features no_std` (it currently just runs with `hashbrown` explicitly), but that seems to pass locally.

File: lightning/src/sync.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+		self.lock.deref_mut()
+	}
+}
+
+impl<T> Mutex<T> {
+	pub fn new(inner: T) -> Mutex<T> {
+		Mutex { inner: RefCell::new(inner) }
+	}
+
+	pub fn lock(&self) -> LockResult<MutexGuard<'_, T>> {
```
Comment (TheBlueMatt) [inline lightning/src/sync.rs:36]: Can you make the self lifetime explicit (and in RwLock)? I know the compiler should do the right thing here but I kinda like taking the reference explicitly.

Comment (devrandom) [reply lightning/src/sync.rs:36]: Do you mean:

```rust
	pub fn lock<'a>(&'a self) -> LockResult<MutexGuard<'a, T>> {
```

**Comment by devrandom** [comment]
> Cool! Glad it kinda just worked. CI should also at least run `cargo test --no-default-features --features no_std` (it currently just runs with `hashbrown` explicitly), but that seems to pass locally.

I read what you said as "instead of `--features hashbrown`", but let me know if I misunderstood.  Will do.

File: lightning/src/sync.rs (hunk 1 above)
Comment (TheBlueMatt) [reply lightning/src/sync.rs:36]: Yes.

Comment (devrandom) [reply lightning/src/sync.rs]: Done

File: lightning/src/sync.rs [hunk 2]
```diff
@@ -34,11 +33,11 @@ impl<T> Mutex<T> {
 		Mutex { inner: RefCell::new(inner) }
 	}
 
-	pub fn lock(&self) -> LockResult<MutexGuard<'_, T>> {
+	pub fn lock<'a>(&'a self) -> LockResult<MutexGuard<'a, T>> {
```
Comment (jkczyz) [inline lightning/src/sync.rs:36]: Could you move the explicit lifetimes into the previous commit?

File: lightning/src/sync.rs [hunk 3]
```diff
@@ -50,18 +49,69 @@ impl Condvar {
...
 		Ok((guard, ()))
 	}
 
 	pub fn notify_all(&self) {}
-}
\ No newline at end of file
+}
+
+/// This no_std version of RwLock actually only allows one lock at a time, because
+/// we are not expecting reentrancy in our straight-line code.
```
Comment (jkczyz) [inline lightning/src/sync.rs:65]: Is this comment about reentrancy still accurate given our Slack discussion on `ChannelManager::process_pending_events`? Maybe should say "one write lock"?

**Review by jkczyz — APPROVED** [review]

File: lightning/src/lib.rs [hunk 4]
```diff
@@ -53,3 +53,7 @@ mod prelude {
 	#[cfg(feature = "hashbrown")]
 	pub use self::hashbrown::{HashMap, HashSet, hash_map};
 }
+
+mod sync {
+	pub use ::std::sync::{Arc, Mutex, Condvar, MutexGuard, RwLock, RwLockReadGuard};
```
[the remark this answers is not in the document]
Comment (devrandom) [reply lightning/src/lib.rs:59]: reordered

File: lightning/src/lib.rs [hunk 5]
```diff
@@ -54,6 +54,10 @@ mod prelude {
 	pub use self::hashbrown::{HashMap, HashSet, hash_map};
 }
 
+#[cfg(feature = "std")]
 mod sync {
 	pub use ::std::sync::{Arc, Mutex, Condvar, MutexGuard, RwLock, RwLockReadGuard};
 }
+
+#[cfg(not(feature = "std"))]
+mod sync;
```
[the remark this answers is not in the document]
Comment (devrandom) [reply lightning/src/lib.rs]: added

File: lightning/src/ln/channel.rs [hunk 6]
```diff
@@ -45,7 +45,8 @@ use prelude::*;
 use core::{cmp,mem,fmt};
 use core::ops::Deref;
 #[cfg(any(test, feature = "fuzztarget"))]
-use std::sync::Mutex;
+use sync::Mutex;
+use sync;
```
[the remark this answers is not in the document]
Comment (devrandom) [reply lightning/src/ln/channel.rs:50]: it's used below in `sync::Mutex`.  but adding a conditional on debug_assertions.

File: lightning/src/sync.rs (hunk 2 above)
Comment (devrandom) [reply lightning/src/sync.rs:55]: squashing into a single commit

File: lightning/src/sync.rs (hunk 3 above)
Comment (devrandom) [reply lightning/src/sync.rs]: yeah, obsolete

File: lightning/src/ln/channel.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (devrandom) [reply lightning/src/ln/channel.rs:50]: actually, it's a bit silly, importing Mutex twice, once for fuzz, and once for debug_assertions.  merging them.

File: lightning/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/lib.rs:59]: Looks like ordering was lost along the way

**Review by jkczyz — APPROVED** [review]
ACK 2e8f4fef2faa06a8752163bb6bc6908520632e27

**Review by TheBlueMatt — APPROVED** [review]

**Pull request merged**